src/v/kafka/server/handlers/alter_partition_reassignments.cc

```diff
@@ -160,28 +160,51 @@ partitions_request_iterator validate_partitions(
           return unkown_broker_id_it == partition.replicas->end();
       });
 
+    // Check for undefined topic here instead of outside
+    // validate_partitions because there are only two places to store errors,
+    // the top level and reassignable_partition. The top level should be for
+    // errors that apply to all topics, not a single topic.
     valid_partitions_end = validate_replicas(
       begin,
       valid_partitions_end,
       std::back_inserter(invalid_partitions),
-      error_code::invalid_replication_factor,
-      "Number of replicas does not match the topic replication factor",
+      error_code::unknown_topic_or_partition,
+      "Topic or partition is undefined",
       [&tp_metadata](const reassignable_partition& partition) {
           if (!partition.replicas.has_value()) {
               return true;
           }
 
-          if (
-            !tp_metadata.has_value()
-            || !tp_metadata.value().is_topic_replicable()) {
-              return false;
-          }
-
-          auto tp_replication_factor
-            = tp_metadata.value().get_replication_factor();
-          return size_t(tp_replication_factor) == partition.replicas->size();
+          return tp_metadata.has_value();
       });
 
+    // Check the replication factor when topic metadata is defined.
+    if (tp_metadata.has_value()) {
+        valid_partitions_end = validate_replicas(
+          begin,
+          valid_partitions_end,
+          std::back_inserter(invalid_partitions),
+          error_code::invalid_replication_factor,
+          "Number of replicas does not match the topic replication factor",
+          [&tp_metadata](const reassignable_partition& partition) {
+              if (!partition.replicas.has_value()) {
```

Comment:
same for here

Reply:
It is necessary here because the actual check for replication factor gets the size of `partition.replicas`. Thus `partition.replicas` must be defined.
```cpp
return size_t(tp_replication_factor) == partition.replicas->size();
```

Reply:
But what should the behavior be if someone makes a request without replicas? From the KIP:
```
INVALID_REPLICA_ASSIGNMENT - when the replica assignment is not valid (brokers are not part of the cluster, negative replica ids, empty assignment)
```
`empty_assignment` is in fact a place where this error code may be returned

Reply:
> But what should the behavior be if someone makes a request without replicas? From the KIP:
> 
> ```
> INVALID_REPLICA_ASSIGNMENT - when the replica assignment is not valid (brokers are not part of the cluster, negative replica ids, empty assignment)
> ```
> 
> `empty_assignment` is in fact a place where this error code may be returned

Those checks are handled in other validators and I have unit tests for them.

For empty assignment:
Validator [here](https://github.com/redpanda-data/redpanda/blob/dev/src/v/kafka/server/handlers/alter_partition_reassignments.cc#L98)
Unit test [here](https://github.com/redpanda-data/redpanda/blob/dev/src/v/kafka/server/tests/partition_reassignments_test.cc#L101)

For brokers not in cluster:
Validator [here](https://github.com/redpanda-data/redpanda/blob/dev/src/v/kafka/server/handlers/alter_partition_reassignments.cc#L146)
Unit test [here](https://github.com/redpanda-data/redpanda/blob/dev/src/v/kafka/server/tests/partition_reassignments_test.cc#L178)

For negative replica ids:
Validator [here](https://github.com/redpanda-data/redpanda/blob/dev/src/v/kafka/server/handlers/alter_partition_reassignments.cc#L128)
Unit test [here](https://github.com/redpanda-data/redpanda/blob/dev/src/v/kafka/server/tests/partition_reassignments_test.cc#L153)

Finally, those validators are ordered before this validator for invalid replication factor. Thus, they will fire first.

Reply:
Maybe I misunderstood your Q.
An empty but not null `replicas` means the user wants to cancel an ongoing reassignment. It is a valid request. So that line is meant to ignore that case.

Reply:
I see , so then we just need to remove the duplicate conditionals which could never occur